File: checker.py
#!/usr/bin/env python3
"""Discord username availability checker.

Checks whether a Discord unique username (the new "@handle" style name,
e.g. "coolname") is currently available or already taken, by calling
Discord's public unauthenticated username-attempt endpoint.
"""

from __future__ import annotations

import argparse
import os
import sys
import time
from concurrent.futures import ThreadPoolExecutor
from typing import Callable, Iterable, List, Optional, Sequence

import requests


API_URL = "https://discord.com/api/v9/unique-username/username-attempt-unauthed"
USER_AGENT = (
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
    "AppleWebKit/537.36 (KHTML, like Gecko) "
    "Chrome/124.0.0.0 Safari/537.36"
)
REQUEST_TIMEOUT = 10
# Default inter-request delay. Kept at 0 so batch checks are fast by default;
# raise it with --delay if Discord starts rate-limiting you.
DEFAULT_DELAY = 0.0
DEFAULT_WORKERS = 8
# Cap how long we'll honor a Retry-After before giving up on the auto-retry.
MAX_RETRY_AFTER_SECONDS = 10.0
# Default file to append results to after a batch check.
DEFAULT_OUTPUT_FILE = "results.txt"
RESULTS_HEADER = "# Discord username check results"

# Status constants. Used both as internal values and (uppercased) in the
# saved results file.
STATUS_AVAILABLE = "available"
STATUS_TAKEN = "taken"
STATUS_INVALID = "invalid"
STATUS_RATE_LIMITED = "rate_limited"
STATUS_ERROR = "error"


class CheckResult:
    """Result of a single username check."""

    def __init__(
        self,
        username: str,
        status: str,
        reason: Optional[str] = None,
    ) -> None:
        self.username = username
        # status is one of: "available", "taken", "invalid", "rate_limited", "error"
        self.status = status
        self.reason = reason

    def format_line(self) -> str:
        if self.status == STATUS_AVAILABLE:
            return f"[+] {self.username}: AVAILABLE"
        if self.status == STATUS_TAKEN:
            return f"[-] {self.username}: taken"
        if self.status == STATUS_INVALID:
            extra = f" ({self.reason})" if self.reason else ""
            return f"[!] {self.username}: invalid{extra}"
        if self.status == STATUS_RATE_LIMITED:
            extra = f" ({self.reason})" if self.reason else ""
            return f"[?] {self.username}: rate limited{extra}"
        # error
        extra = f" ({self.reason})" if self.reason else ""
        return f"[?] {self.username}: error{extra}"

    def format_file_line(self) -> str:
        """Render this result as a line for the results file.

        Format: ``<username> - <STATUS>`` with optional ``: <reason>`` for
        statuses that carry extra detail (invalid / error / rate_limited).
        """
        # rate_limited reads better as "RATE_LIMITED" in the file.
        status_token = self.status.upper()
        if self.status in (STATUS_INVALID, STATUS_ERROR, STATUS_RATE_LIMITED) and self.reason:
            return f"{self.username} - {status_token}: {self.reason}"
        return f"{self.username} - {status_token}"


def _post_attempt(
    username: str, session: requests.Session
) -> requests.Response:
    return session.post(
        API_URL,
        json={"username": username},
        headers={
            "User-Agent": USER_AGENT,
            "Accept": "application/json",
            "Content-Type": "application/json",
        },
        timeout=REQUEST_TIMEOUT,
    )


def _retry_after_seconds(response: requests.Response) -> Optional[float]:
    """Extract a Retry-After value (seconds) from a 429 response, if any."""
    header = response.headers.get("Retry-After")
    if header:
        try:
            return float(header)
        except ValueError:
            pass
    try:
        body = response.json()
    except ValueError:
        return None
    if isinstance(body, dict):
        value = body.get("retry_after")
        if isinstance(value, (int, float)):
            return float(value)
    return None


def check_username(username: str, session: requests.Session) -> CheckResult:
    """Check a single username against Discord's API.

    Returns a CheckResult and never raises for normal failures. On HTTP 429
    the call will sleep up to MAX_RETRY_AFTER_SECONDS and retry once.
    """
    username = username.strip()
    if not username:
        return CheckResult(username, STATUS_INVALID, "empty username")

    try:
        response = _post_attempt(username, session)
    except requests.RequestException as exc:
        return CheckResult(username, STATUS_ERROR, f"network error: {exc}")

    if response.status_code == 429:
        wait = _retry_after_seconds(response)
        if wait is not None and wait <= MAX_RETRY_AFTER_SECONDS:
            time.sleep(max(wait, 0.0))
            try:
                response = _post_attempt(username, session)
            except requests.RequestException as exc:
                return CheckResult(username, STATUS_ERROR, f"network error: {exc}")
            if response.status_code == 429:
                wait2 = _retry_after_seconds(response)
                detail = f"retry after {wait2:.2f}s" if wait2 is not None else None
                return CheckResult(username, STATUS_RATE_LIMITED, detail)
        else:
            detail = (
                f"retry after {wait:.2f}s" if wait is not None else None
            )
            return CheckResult(username, STATUS_RATE_LIMITED, detail)

    # Try to parse JSON regardless of status code, since Discord returns
    # validation errors as JSON with non-2xx codes too.
    try:
        data = response.json()
    except ValueError:
        return CheckResult(
            username,
            STATUS_ERROR,
            f"HTTP {response.status_code}, non-JSON response",
        )

    if response.status_code == 200 and isinstance(data, dict) and "taken" in data:
        return CheckResult(
            username,
            STATUS_TAKEN if data["taken"] else STATUS_AVAILABLE,
        )

    # Validation / rate-limit / other error cases.
    reason = _extract_error_reason(data) or f"HTTP {response.status_code}"
    # Treat clear validation errors on the username field as "invalid",
    # everything else as a generic error.
    if response.status_code in (400, 422) and isinstance(data, dict):
        return CheckResult(username, STATUS_INVALID, reason)
    return CheckResult(username, STATUS_ERROR, reason)


def _extract_error_reason(data: object) -> Optional[str]:
    """Pull a human-readable reason out of a Discord error JSON payload."""
    if not isinstance(data, dict):
        return None

    # Common top-level message field.
    message = data.get("message")

    # Field-specific errors live under "errors": {"username": {"_errors": [...]}}.
    field_messages = []
    errors = data.get("errors")
    if isinstance(errors, dict):
        for field, value in errors.items():
            if isinstance(value, dict) and "_errors" in value:
                for entry in value["_errors"]:
                    if isinstance(entry, dict) and entry.get("message"):
                        field_messages.append(f"{field}: {entry['message']}")

    parts = []
    if field_messages:
        parts.extend(field_messages)
    elif message:
        parts.append(str(message))

    return "; ".join(parts) if parts else None


def save_results(results: Iterable[CheckResult], path: str) -> int:
    """Append a batch of results to ``path`` in the standard text format.

    If the file does not exist yet, it is created and a header comment line
    (``RESULTS_HEADER``) is written first. Subsequent runs against the same
    file just append more lines without repeating the header.

    Returns the number of result lines written.
    """
    results_list = list(results)
    if not results_list:
        return 0

    file_exists = os.path.exists(path)
    with open(path, "a", encoding="utf-8") as fh:
        if not file_exists:
            fh.write(RESULTS_HEADER + "\n")
        for result in results_list:
            fh.write(result.format_file_line() + "\n")
    return len(results_list)


def parse_usernames_text(text: str) -> List[str]:
    """Parse a blob of user-supplied text into a list of usernames.

    Accepts newline-separated, comma-separated, or whitespace-separated
    input and any mix of those. Empty fragments are dropped. Order is
    preserved.
    """
    if not text:
        return []
    # Replace commas with newlines, then split on any whitespace.
    pieces = text.replace(",", "\n").split()
    return [p.strip() for p in pieces if p.strip()]


def load_usernames_from_file(path: str) -> List[str]:
    """Read usernames from a text file (one per line or comma-separated)."""
    with open(path, "r", encoding="utf-8") as fh:
        return parse_usernames_text(fh.read())


def check_many(
    usernames: Iterable[str],
    delay: float = DEFAULT_DELAY,
    workers: int = DEFAULT_WORKERS,
    session: Optional[requests.Session] = None,
    on_result: Optional[Callable[[int, CheckResult], None]] = None,
) -> List[CheckResult]:
    """Check several usernames concurrently and return results in input order.

    A single shared ``requests.Session`` is reused across worker threads.
    If ``delay`` is greater than zero, each task sleeps for that many seconds
    before its request, which throttles the overall request rate when
    combined with a small worker count.

    ``on_result`` is invoked from worker threads as each result completes,
    receiving ``(index, CheckResult)``. The GUI uses this to stream progress.
    Results in the returned list are always ordered to match the input.
    """
    names: List[str] = [str(n) for n in usernames]
    if not names:
        return []

    owns_session = session is None
    if session is None:
        session = requests.Session()
    effective_workers = max(1, min(workers, len(names)))

    results: List[Optional[CheckResult]] = [None] * len(names)

    def _task(index_name):
        index, name = index_name
        if delay > 0 and index > 0:
            time.sleep(delay)
        result = check_username(name, session)
        results[index] = result
        if on_result is not None:
            try:
                on_result(index, result)
            except Exception:
                # Never let a callback failure break the batch.
                pass
        return result

    try:
        if effective_workers == 1:
            for pair in enumerate(names):
                _task(pair)
        else:
            with ThreadPoolExecutor(max_workers=effective_workers) as pool:
                # Drain to make sure any exceptions surface.
                for _ in pool.map(_task, list(enumerate(names))):
                    pass
    finally:
        if owns_session:
            session.close()

    # mypy/runtime: by this point every slot is filled.
    return [r for r in results if r is not None]


def run_batch(
    usernames: Sequence[str],
    delay: float = DEFAULT_DELAY,
    workers: int = DEFAULT_WORKERS,
    output_file: Optional[str] = DEFAULT_OUTPUT_FILE,
) -> int:
    """High-level CLI batch: check, print, and optionally save.

    Returns a process exit code (2 if any check errored).
    """
    if not usernames:
        return 0

    results = check_many(usernames, delay=delay, workers=workers)

    exit_code = 0
    for result in results:
        print(result.format_line())
        if result.status == STATUS_ERROR:
            exit_code = 2

    if output_file:
        try:
            written = save_results(results, output_file)
        except OSError as exc:
            print(
                f"warning: could not write to {output_file}: {exc}",
                file=sys.stderr,
            )
        else:
            if written:
                print(f"\nSaved {written} result(s) to {output_file}")

    return exit_code


def interactive_loop() -> int:
    """Prompt the user for usernames until they quit."""
    print("Discord username availability checker")
    print("Type a username to check, or 'quit' / 'exit' to stop.")
    session = requests.Session()
    try:
        while True:
            try:
                raw = input("> ")
            except EOFError:
                print()
                return 0
            name = raw.strip()
            if not name:
                continue
            if name.lower() in ("quit", "exit"):
                return 0
            result = check_username(name, session)
            print(result.format_line())
    except KeyboardInterrupt:
        print()
        return 0


def parse_args(argv=None) -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description=(
            "Check whether one or more Discord unique usernames are "
            "available or already taken."
        ),
    )
    parser.add_argument(
        "usernames",
        nargs="*",
        help="Usernames to check. If omitted (and --input is not used), runs in interactive mode.",
    )
    parser.add_argument(
        "--input",
        "-i",
        metavar="PATH",
        help=(
            "Read usernames from a text file (one per line or "
            "comma-separated). Combined with any usernames given on the "
            "command line."
        ),
    )
    parser.add_argument(
        "--delay",
        type=float,
        default=DEFAULT_DELAY,
        help=(
            "Seconds to wait before each request when checking multiple "
            "names. Use this if Discord starts rate-limiting you "
            "(default: %(default)s)."
        ),
    )
    parser.add_argument(
        "--workers",
        type=int,
        default=DEFAULT_WORKERS,
        help=(
            "Number of concurrent worker threads for batch checks "
            "(default: %(default)s). Use 1 to force sequential checks."
        ),
    )
    parser.add_argument(
        "--output",
        "-o",
        default=DEFAULT_OUTPUT_FILE,
        help=(
            "After a batch check, append all results to this file, "
            "creating it (with a header) if needed "
            "(default: %(default)s)."
        ),
    )
    parser.add_argument(
        "--no-save",
        dest="output",
        action="store_const",
        const=None,
        help="Don't write results to a file.",
    )
    return parser.parse_args(argv)


def main(argv=None) -> int:
    args = parse_args(argv)
    if args.workers < 1:
        print("error: --workers must be >= 1", file=sys.stderr)
        return 2

    usernames: List[str] = list(args.usernames)
    if args.input:
        try:
            usernames.extend(load_usernames_from_file(args.input))
        except OSError as exc:
            print(f"error: could not read --input file: {exc}", file=sys.stderr)
            return 2

    if usernames:
        return run_batch(
            usernames,
            delay=args.delay,
            workers=args.workers,
            output_file=args.output,
        )
    return interactive_loop()


if __name__ == "__main__":
    sys.exit(main())

File: src/checker.rs
//! Async HTTP client for Discord's public username-attempt endpoint.
//!
//! The endpoint is unauthenticated and is what the Discord web client
//! itself calls while a user is picking a new handle:
//!
//! ```text
//! POST https://discord.com/api/v9/unique-username/username-attempt-unauthed
//! Content-Type: application/json
//! {"username": "<name>"}
//! ```
//!
//! On 200 the response body is `{"taken": true|false}`. Validation
//! errors come back as 400/422 with a JSON body; rate limiting as 429
//! with an optional `Retry-After` header / `retry_after` body field.
//!
//! Errors are folded into [`Status`] variants — `check_username` never
//! returns `Err` for ordinary failures so batch loops keep flowing.

use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::time::{Duration, Instant};

use reqwest::{Client, StatusCode};
use serde::Deserialize;
use tokio::sync::Semaphore;

use crate::result::{CheckResult, Status};

const API_URL: &str = "https://discord.com/api/v9/unique-username/username-attempt-unauthed";

const USER_AGENT: &str = concat!(
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) ",
    "AppleWebKit/537.36 (KHTML, like Gecko) ",
    "Chrome/124.0.0.0 Safari/537.36"
);

const REQUEST_TIMEOUT: Duration = Duration::from_secs(10);
const MAX_AUTO_RETRY_AFTER: Duration = Duration::from_secs(10);

/// Tunables for a [`Checker`] / batch run.
#[derive(Debug, Clone)]
pub struct CheckerConfig {
    /// Maximum simultaneous in-flight HTTP requests.
    pub workers: usize,
    /// Minimum delay between task starts. Helpful for polite throttling.
    pub delay: Duration,
    /// Per-request timeout for HTTP I/O.
    pub timeout: Duration,
    /// If a 429 quotes a Retry-After shorter than this, auto-retry once.
    pub max_auto_retry_after: Duration,
}

impl Default for CheckerConfig {
    fn default() -> Self {
        Self {
            workers: 8,
            delay: Duration::from_millis(0),
            timeout: REQUEST_TIMEOUT,
            max_auto_retry_after: MAX_AUTO_RETRY_AFTER,
        }
    }
}

/// Streaming callback invoked from worker tasks as results arrive.
///
/// The order of invocations is non-deterministic; the final aggregated
/// `Vec<CheckResult>` returned by [`Checker::run_batch`] always matches
/// the input order.
pub type BatchProgress = Arc<dyn Fn(usize, &CheckResult) + Send + Sync>;

/// Reusable async HTTP client wrapper.
#[derive(Clone)]
pub struct Checker {
    client: Client,
    config: CheckerConfig,
}

impl Checker {
    pub fn new(config: CheckerConfig) -> reqwest::Result<Self> {
        let client = Client::builder()
            .user_agent(USER_AGENT)
            .timeout(config.timeout)
            .pool_idle_timeout(Some(Duration::from_secs(30)))
            .build()?;
        Ok(Self { client, config })
    }

    pub fn config(&self) -> &CheckerConfig {
        &self.config
    }

    /// Check a single username. Never panics; failures are encoded as
    /// `Status::Error` / `Status::Invalid` / `Status::RateLimited`.
    pub async fn check(&self, username: &str) -> CheckResult {
        let name = username.trim();
        if name.is_empty() {
            return CheckResult::new(
                "",
                Status::Invalid {
                    reason: "empty username".into(),
                },
            );
        }

        match self.do_attempt(name).await {
            Ok(res) => res,
            Err(err) => CheckResult::new(
                name,
                Status::Error {
                    reason: format!("network error: {err}"),
                },
            ),
        }
    }

    async fn do_attempt(&self, username: &str) -> reqwest::Result<CheckResult> {
        let resp = self.post_attempt(username).await?;

        if resp.status() == StatusCode::TOO_MANY_REQUESTS {
            let wait = parse_retry_after(&resp).await;
            match wait {
                Some(d) if d <= self.config.max_auto_retry_after => {
                    tokio::time::sleep(d).await;
                    let retry = self.post_attempt(username).await?;
                    if retry.status() == StatusCode::TOO_MANY_REQUESTS {
                        let again = parse_retry_after(&retry).await;
                        return Ok(CheckResult::new(
                            username,
                            Status::RateLimited {
                                retry_after_ms: again.map(|d| d.as_millis() as u64),
                            },
                        ));
                    }
                    return interpret_response(username, retry).await;
                }
                Some(d) => {
                    return Ok(CheckResult::new(
                        username,
                        Status::RateLimited {
                            retry_after_ms: Some(d.as_millis() as u64),
                        },
                    ));
                }
                None => {
                    return Ok(CheckResult::new(
                        username,
                        Status::RateLimited {
                            retry_after_ms: None,
                        },
                    ));
                }
            }
        }

        interpret_response(username, resp).await
    }

    async fn post_attempt(&self, username: &str) -> reqwest::Result<reqwest::Response> {
        self.client
            .post(API_URL)
            .json(&serde_json::json!({ "username": username }))
            .header("Accept", "application/json")
            .send()
            .await
    }

    /// Run a concurrent batch of checks. Results come back in input
    /// order. `progress` is called from worker tasks for each completed
    /// item — keep the callback cheap and non-blocking.
    ///
    /// If `cancel` is provided and flips to `true`, any task that has
    /// not yet started will short-circuit; in-flight tasks finish
    /// normally. The returned `Vec` is truncated to only those slots
    /// that produced a result.
    pub async fn run_batch(
        &self,
        usernames: Vec<String>,
        progress: Option<BatchProgress>,
        cancel: Option<Arc<AtomicBool>>,
    ) -> Vec<CheckResult> {
        if usernames.is_empty() {
            return Vec::new();
        }

        let workers = self.config.workers.max(1).min(usernames.len());
        let semaphore = Arc::new(Semaphore::new(workers));
        let delay = self.config.delay;
        let start = Instant::now();

        let mut handles = Vec::with_capacity(usernames.len());
        for (idx, name) in usernames.iter().enumerate() {
            let checker = self.clone();
            let name = name.clone();
            let sem = semaphore.clone();
            let progress = progress.clone();
            let cancel = cancel.clone();
            handles.push(tokio::spawn(async move {
                if let Some(c) = &cancel {
                    if c.load(Ordering::SeqCst) {
                        return (idx, None);
                    }
                }
                let _permit = sem.acquire_owned().await.expect("semaphore closed");
                if let Some(c) = &cancel {
                    if c.load(Ordering::SeqCst) {
                        return (idx, None);
                    }
                }
                if idx > 0 && delay > Duration::ZERO {
                    // Stagger by `delay * idx` from batch start, capped so a
                    // long batch doesn't drift unboundedly.
                    let target = start + delay.saturating_mul(idx.min(1024) as u32);
                    let now = Instant::now();
                    if target > now {
                        tokio::time::sleep(target - now).await;
                    }
                }
                let result = checker.check(&name).await;
                if let Some(cb) = progress {
                    cb(idx, &result);
                }
                (idx, Some(result))
            }));
        }

        let mut bucket: Vec<Option<CheckResult>> = (0..usernames.len()).map(|_| None).collect();
        for h in handles {
            if let Ok((idx, res)) = h.await {
                if let (Some(r), Some(slot)) = (res, bucket.get_mut(idx)) {
                    *slot = Some(r);
                }
            }
        }
        bucket.into_iter().flatten().collect()
    }
}

async fn parse_retry_after(resp: &reqwest::Response) -> Option<Duration> {
    if let Some(header) = resp.headers().get(reqwest::header::RETRY_AFTER) {
        if let Ok(s) = header.to_str() {
            if let Ok(secs) = s.parse::<f64>() {
                return Some(Duration::from_millis((secs.max(0.0) * 1000.0) as u64));
            }
        }
    }
    // Falling through to body parsing requires consuming the response,
    // which the caller still wants to read. So we only consult the
    // header here; body-side retry_after is parsed later in
    // `interpret_response` when we already own the body.
    None
}

async fn interpret_response(
    username: &str,
    resp: reqwest::Response,
) -> reqwest::Result<CheckResult> {
    let status_code = resp.status();
    let text = resp.text().await?;
    let parsed: Option<serde_json::Value> = serde_json::from_str(&text).ok();

    if status_code == StatusCode::OK {
        if let Some(value) = &parsed {
            if let Some(taken) = value.get("taken").and_then(|v| v.as_bool()) {
                let status = if taken {
                    Status::Taken
                } else {
                    Status::Available
                };
                return Ok(CheckResult::new(username, status));
            }
        }
        return Ok(CheckResult::new(
            username,
            Status::Error {
                reason: format!("HTTP 200, unexpected body: {}", clip(&text, 200)),
            },
        ));
    }

    if status_code == StatusCode::TOO_MANY_REQUESTS {
        let ms = parsed
            .as_ref()
            .and_then(|v| v.get("retry_after"))
            .and_then(|v| v.as_f64())
            .map(|s| (s.max(0.0) * 1000.0) as u64);
        return Ok(CheckResult::new(
            username,
            Status::RateLimited { retry_after_ms: ms },
        ));
    }

    let reason = parsed
        .as_ref()
        .and_then(extract_error_reason)
        .unwrap_or_else(|| format!("HTTP {}", status_code.as_u16()));

    let status = if matches!(status_code.as_u16(), 400 | 422) {
        Status::Invalid { reason }
    } else {
        Status::Error { reason }
    };
    Ok(CheckResult::new(username, status))
}

fn clip(s: &str, n: usize) -> String {
    if s.chars().count() <= n {
        s.to_string()
    } else {
        let mut out: String = s.chars().take(n).collect();
        out.push('…');
        out
    }
}

#[derive(Deserialize)]
struct DiscordFieldError {
    message: Option<String>,
}

fn extract_error_reason(value: &serde_json::Value) -> Option<String> {
    if !value.is_object() {
        return None;
    }

    // Field-specific: errors.<field>._errors[].message
    let mut field_messages: Vec<String> = Vec::new();
    if let Some(errors) = value.get("errors").and_then(|v| v.as_object()) {
        for (field, payload) in errors {
            if let Some(inner) = payload.get("_errors").and_then(|v| v.as_array()) {
                for entry in inner {
                    if let Ok(parsed) = serde_json::from_value::<DiscordFieldError>(entry.clone()) {
                        if let Some(msg) = parsed.message {
                            field_messages.push(format!("{field}: {msg}"));
                        }
                    }
                }
            }
        }
    }
    if !field_messages.is_empty() {
        return Some(field_messages.join("; "));
    }

    // Top-level message fallback.
    value
        .get("message")
        .and_then(|v| v.as_str())
        .map(|s| s.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn extracts_field_error() {
        let json = serde_json::json!({
            "errors": {
                "username": {
                    "_errors": [
                        {"code": "USERNAME_TOO_SHORT", "message": "Must be 2-32 chars."}
                    ]
                }
            }
        });
        let reason = extract_error_reason(&json).unwrap();
        assert!(reason.contains("username"));
        assert!(reason.contains("Must be 2-32 chars."));
    }

    #[test]
    fn extracts_top_level_message() {
        let json = serde_json::json!({ "message": "You are being rate limited" });
        let reason = extract_error_reason(&json).unwrap();
        assert_eq!(reason, "You are being rate limited");
    }

    #[test]
    fn clip_truncates() {
        assert_eq!(clip("abcdef", 3), "abc…");
        assert_eq!(clip("abc", 10), "abc");
    }
}
